HPCC-19168 A IFile::size() call for a dafilesrv file throws an exception (8019) if file is missing, a std. (local) file returns (offset_t)-1

AttilaVamos opened the pull request:
Add code to handle RFSERR_GetDirFailed (8019) exception in IFile::size() method 
and return with -1 if remote file is missing.

Signed-off-by: Attila Vamos <attila.vamos@gmail.com>

<!-- Thank you for submitting a pull request to the HPCC project

 PLEASE READ the following before proceeding.

 This project only accepts pull requests related to open JIRA issues.
 If suggesting a new feature or change, please discuss it in a JIRA issue first.
 If fixing a bug, there should be an issue describing it with steps to reproduce.
 The title line of the pull request (and of each commit within it) should refer to the
 associated issue using the format:

 HPCC-nnnnn Short description of issue

 This will allow the Jira ticket to be automatically updated to refer to this pull request,
 and will ensure that the automatically-generated changelog is properly formatted.
 Where a pull request contains a single commit the pull request title will be set automatically,
 assuming that the commit has followed the proper guidelines.

 Please go over all the following points, and put an `x` in all the boxes that apply. You may find
 it easier to press the 'Create' button first then click on the checkboxes to edit the comment.
-->

## Type of change:
- [x] This change is a bug fix (non-breaking change which fixes an issue).
- [ ] This change is a new feature (non-breaking change which adds functionality).
- [ ] This change improves the code (refactor or other change that does not change the functionality)
- [ ] This change fixes warnings (the fix does not alter the functionality or the generated code)
- [ ] This change is a breaking change (fix or feature that will cause existing behavior to change).
- [ ] This change alters the query API (existing queries will have to be recompiled)

## Checklist:
- [x] My code follows the code style of this project.
  - [x] My code does not create any new warnings from compiler, build system, or lint.
- [x] The commit message is properly formatted and free of typos.
  - [x] The commit message title makes sense in a changelog, by itself.
  - [x] The commit is signed.
- [ ] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly, or...
  - [ ] I have created a JIRA ticket to update the documentation.
  - [ ] Any new interfaces or exported functions are appropriately commented.
- [x] I have read the CONTRIBUTORS document.
- [x] The change has been fully tested:
  - [ ] I have added tests to cover my changes.
  - [x] All new and existing tests passed.
  - [ ] I have checked that this change does not introduce memory leaks.
  - [ ] I have used Valgrind or similar tools to check for potential issues.
- [ ] I have given due consideration to all of the following potential concerns:
  - [ ] Scalability
  - [ ] Performance
  - [ ] Security
  - [ ] Thread-safety
  - [ ] Premature optimization
  - [x] Existing deployed queries will not be broken
  - [x] This change fixes the problem, not just the symptom
  - [x] The target branch of this pull request is appropriate for such a change.
- [x] There are no similar instances of the same problem that should be addressed
  - [ ] I have addressed them here
  - [ ] I have raised JIRA issues to address them separately
- [ ] This is a user interface / front-end modification
  - [ ] I have tested my changes in multiple modern browsers
  - [ ] The component(s) render as expected

## Testing:
<!-- Please describe how this change has been tested.-->
Tested manually on a 2 nodes 4 slaves system.
<!-- Thank you for taking the time to submit this pull request and to answer all of the above-->

jakesmith commented on common/remote/sockfile.cpp:
```diff
@@ -1269,8 +1269,10 @@ protected: friend class CRemoteFileIO;
-            else if(errCode == 8209)
+            else if (errCode == 8209)
                 msg.append("Failed to open directory.");
+            else if (errCode == RFSERR_GetDirFailed)
```
Lines 1272 and 1273 should be deleted, otherwise this line can't be reached.

AttilaVamos replied:
Why? The RFSERR_GetDirFailed==8019 (as you can see in remoteerr.hpp) and not 8209. There are some multiple local (re)definition for RFSERR_GetDirFailed macro. I think I will open a JIRA to make one global definition for it.

I left that check for compatibility consideration (if some dafilesrv uses 8209 value instead of 8019.) 
Perhaps I should add the 8209 check to the size as well.

jakesmith replied:
hm, multiple different definitions of RFSERR_GetDirFailed, yuk - yes I can see them now.

in rfs.cpp
```
#define RFSERR_GetDirFailed                     8019
```

and in daliservix.cpp
```
#define RFSERR_GetDirFailed                     8209
```

The daliservix is supposed to be a solaris version of dafilesrv.
Why they have different codes... I do not know, but best leave it as is then.

jakesmith commented on common/remote/sockfile.cpp:
```diff
@@ -1964,7 +1966,20 @@ class CRemoteFile : public CRemoteBase, implements IFile
+            {
+                sendRemoteCommand(sendBuffer, replyBuffer);
+            }
+            catch (IException * e)
```
More accurate to catch IDAFS_Exception here.

AttilaVamos replied:
Changed

jakesmith commented:
@AttilaVamos - a couple of minor comments, but looks good.

HPCCSmoketest commented:
Automated Smoketest: :white_check_mark:
OS: centos 7.2.1511 (Linux 3.10.0-327.28.3.el7.x86_64)
Sha: bdef9e1eb5291e7c2e8f52271a007688a897cec9
Build: success
Install hpccsystems-platform-community_6.4.14-closedown0.el7.x86_64.rpm
HPCC Start: OK

Unit tests result:

| Test | total | passed | failed | errors | timeout | 
|---|---|---|---|---|---|
|unittest |97 |97 |0 |0 |0 |
|wutoolTest(Dali) |19 |19 |0 |0 |0 |
|wutoolTest(Cassandra) |19 |19 |0 |0 |0 |



Regression test result:

| phase | total | pass | fail | 
|---|---|---|---|
|setup (hthor) |11 |11 |0 |
|setup (thor) |11 |11 |0 |
|setup (roxie) |11 |11 |0 |
|test (hthor) |734 |734 |0 |
|test (thor) |628 |628 |0 |
|test (roxie) |761 |761 |0 |


HPCC Stop: OK
HPCC Uninstall: OK
Time stats:

| Prep time | Build time | Package time | Install time | Start time | Test time | Stop time | Summary | 
|---|---|---|---|---|---|---|---|
|   14 sec (00:00:14) |  160 sec (00:02:40) |   37 sec (00:00:37) |    6 sec (00:00:06) |   35 sec (00:00:35) | 1123 sec (00:18:43) |   34 sec (00:00:34) | 1409 sec (00:23:29) |

AttilaVamos commented:
@jakesmith Updated, please re-review.

jakesmith commented on common/remote/sockfile.cpp:
```diff
@@ -1964,7 +1967,21 @@ class CRemoteFile : public CRemoteBase, implements IFile
+            }
+            catch (IDAFS_Exception * e)
+            {
+                //TODO This magic number 8209 should clarify and use an RFSERR_xxx macro instead
```
trivial: Probably better to reference that it's the Solaris daliservix.cpp's definition of RFSERR_GetDirFailed and it is unknown why constant is different.

jakesmith commented:
@AttilaVamos - a comment re. comment, but could be merge as is.

richardkchapman commented:
@AttilaVamos Please squash

AttilaVamos committed:
HPCC-19168 A IFile::size() call for a dafilesrv file throws an exception (8019) if file is missing, a std. (local) file returns (offset_t)-1

Add code to handle RFSERR_GetDirFailed (8019) exeption and return with -1
if remote file is missing.

Tested manually on a 2 nodes 4 slaves system.

Signed-off-by: Attila Vamos <attila.vamos@gmail.com>

HPCCSmoketest commented:
Automated Smoketest: :white_check_mark:
OS: centos 7.2.1511 (Linux 3.10.0-327.28.3.el7.x86_64)
Sha: e4a39006a5b4bd443a32cca4c309b4854dcec240
Build: success
Install hpccsystems-platform-community_6.4.14-closedown0.el7.x86_64.rpm
HPCC Start: OK

Unit tests result:

| Test | total | passed | failed | errors | timeout | 
|---|---|---|---|---|---|
|unittest |97 |97 |0 |0 |0 |
|wutoolTest(Dali) |19 |19 |0 |0 |0 |
|wutoolTest(Cassandra) |19 |19 |0 |0 |0 |



Regression test result:

| phase | total | pass | fail | 
|---|---|---|---|
|setup (hthor) |11 |11 |0 |
|setup (thor) |11 |11 |0 |
|setup (roxie) |11 |11 |0 |
|test (hthor) |734 |734 |0 |
|test (thor) |628 |628 |0 |
|test (roxie) |761 |761 |0 |


HPCC Stop: OK
HPCC Uninstall: OK
Time stats:

| Prep time | Build time | Package time | Install time | Start time | Test time | Stop time | Summary | 
|---|---|---|---|---|---|---|---|
|    7 sec (00:00:07) |  159 sec (00:02:39) |   37 sec (00:00:37) |    5 sec (00:00:05) |   33 sec (00:00:33) | 1074 sec (00:17:54) |   33 sec (00:00:33) | 1348 sec (00:22:28) |

AttilaVamos commented:
@richardkchapman squashed and tested.

HPCCSmoketest commented:
Automated Smoketest: :x:
OS: centos 7.2.1511 (Linux 3.10.0-327.28.3.el7.x86_64)
Sha: 91e48d0401718c7b883fcb935ef816ba30ab9daf
Build: success
Install hpccsystems-platform-community_6.4.14-closedown0.el7.x86_64.rpm
HPCC Start: OK

Unit tests result:

| Test | total | passed | failed | errors | timeout | 
|---|---|---|---|---|---|
|unittest |97 |97 |0 |0 |0 |
|wutoolTest(Dali) |19 |19 |0 |0 |0 |
|wutoolTest(Cassandra) |19 |19 |0 |0 |0 |



Regression test result:

| phase | total | pass | fail | 
|---|---|---|---|
|setup (hthor) |11 |11 |0 |
|setup (thor) |11 |11 |0 |
|setup (roxie) |11 |11 |0 |
|test (hthor) |734 |734 |0 |
|test (thor) |628 |628 |0 |
|test (roxie) |761 |760 |1 |


[Errors:](https://gist.github.com/936c1d5400fad6b1f438b893e7d3393d)
- test
  - roxie
    - 525\. [schedule2](https://gist.github.com/HPCCSmoketest/936c1d5400fad6b1f438b893e7d3393d#file-schedule2-w20180301-144653-3-log), 

HPCC Stop: OK
HPCC Uninstall: OK
Time stats:

| Prep time | Build time | Package time | Install time | Start time | Test time | Stop time | Summary | 
|---|---|---|---|---|---|---|---|
|   20 sec (00:00:20) |  158 sec (00:02:38) |   37 sec (00:00:37) |    6 sec (00:00:06) |   36 sec (00:00:36) | 2062 sec (00:34:22) |   32 sec (00:00:32) | 2351 sec (00:39:11) |

HPCCSmoketest commented:
Automated Smoketest: :white_check_mark:
OS: centos 7.2.1511 (Linux 3.10.0-327.28.3.el7.x86_64)
Sha: 91e48d0401718c7b883fcb935ef816ba30ab9daf
Build: success
Install hpccsystems-platform-community_6.4.14-closedown0.el7.x86_64.rpm
HPCC Start: OK

Unit tests result:

| Test | total | passed | failed | errors | timeout | 
|---|---|---|---|---|---|
|unittest |97 |97 |0 |0 |0 |
|wutoolTest(Dali) |19 |19 |0 |0 |0 |
|wutoolTest(Cassandra) |19 |19 |0 |0 |0 |



Regression test result:

| phase | total | pass | fail | 
|---|---|---|---|
|setup (hthor) |11 |11 |0 |
|setup (thor) |11 |11 |0 |
|setup (roxie) |11 |11 |0 |
|test (hthor) |734 |734 |0 |
|test (thor) |628 |628 |0 |
|test (roxie) |761 |761 |0 |


HPCC Stop: OK
HPCC Uninstall: OK
Time stats:

| Prep time | Build time | Package time | Install time | Start time | Test time | Stop time | Summary | 
|---|---|---|---|---|---|---|---|
|    5 sec (00:00:05) |  159 sec (00:02:39) |   37 sec (00:00:37) |    6 sec (00:00:06) |   34 sec (00:00:34) | 1088 sec (00:18:08) |   32 sec (00:00:32) | 1361 sec (00:22:41) |

richardkchapman commented:
@jakesmith Is this a regression? I know there were some changes in this area recently?

jakesmith commented:
Yes looks like it is, testing with 5.6.8 passes.
Broke by HPCC-18725

It's little odd though, it's because IFile::directoryFiles never returns an error. If given an invalid path, it returns a null iterator.
The code in dafilesrv for size() used check if no iterator and return an error - but it would never happen.
The new code (since HPCC-18725) checks to see if the file exists and returns an error if it doesn't.

jakesmith commented:
NB: I think the changes in this PR are good, protecting client side from server exceptions and ensuring standard file size() semantics (i.e. return (offset_t)-1 if file not found).

But getDir (server-side) that's used by things other than size() needs fixing also to address this regression.
I'll push a additional commit in a moment.
